libyul/optimiser/UnusedPruner.cpp

```diff
@@ -120,26 +130,31 @@ void UnusedPruner::runUntilStabilised(
 	Dialect const& _dialect,
 	Block& _ast,
 	bool _allowMSizeOptimization,
+	map<YulString, SideEffects> const* _functionSideEffects,
 	set<YulString> const& _externallyUsedFunctions
 )
 {
 	while (true)
 	{
-		UnusedPruner pruner(_dialect, _ast, _allowMSizeOptimization, _externallyUsedFunctions);
+		UnusedPruner pruner(
+			_dialect, _ast, _allowMSizeOptimization, _functionSideEffects,
+							_externallyUsedFunctions);
 		pruner(_ast);
 		if (!pruner.shouldRunAgain())
 			return;
 	}
 }
 
-void UnusedPruner::runUntilStabilised(
+void UnusedPruner::runUntilStabilisedOnFullAST(
 	Dialect const& _dialect,
 	Block& _ast,
 	set<YulString> const& _externallyUsedFunctions
 )
 {
+	map<YulString, SideEffects> functionSideEffects =
+		SideEffectsPropagator::sideEffects(_dialect, CallGraphGenerator::callGraph(_ast));
 	bool allowMSizeOptimization = !MSizeFinder::containsMSize(_dialect, _ast);
-	runUntilStabilised(_dialect, _ast, allowMSizeOptimization, _externallyUsedFunctions);
+	runUntilStabilised(_dialect, _ast, allowMSizeOptimization, &functionSideEffects, _externallyUsedFunctions);
```

Comment:
Here as well - why not move down by value and default-initifalize as empty map in the constructor?

Reply:
I guess the thought was that we might re-use the map for multiple stages with re-computing, but it turned out that this is not the case.

Reply:
Yeah, we can keep it as it is as well - no optimizer step should ever make side effects of any function *worse*, so it'd always be safe to re-use the map...

Reply:
I'll keep it like that here, because it will be called in a loop.

Reply:
Ah right - I missed the loop, right.